server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java

```diff
@@ -0,0 +1,157 @@
+        DiscoveryNode node = clusterState.nodes().get(nodeId);
+        WeightedRoutingMetadata weightedRoutingMetadata = clusterState.metadata().weightedRoutingMetadata();
+        if (weightedRoutingMetadata != null) {
+            WeightedRouting weightedRouting = weightedRoutingMetadata.getWeightedRouting();
+            // TODO: check weighted routing has weights set after merging versioning changes
+            if (weightedRouting != null) {
+                // Fetch weighted routing attributes with weight set as zero
+                Stream<String> keys = weightedRouting.weights()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().intValue() == 0)
+                    .map(Map.Entry::getKey);
+
+                for (Object key : keys.toArray()) {
+                    if (node.getAttributes().get(weightedRouting.attributeName()).equals(key.toString())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This function returns next shard copy to retry search request in case of failure from previous copy returned
+     * by the iterator. It has the logic to fail open ie request shard copies present in nodes with weighted shard
+     * routing weight set to zero
+     *
+     * @param shardIt Shard Iterator containing order in which shard copies for a shard need to be requested
+     * @return the next shard copy
+     */
+    public SearchShardTarget findNext(final SearchShardIterator shardIt) {
+        SearchShardTarget next = shardIt.nextOrNull();
+        while (next != null && shardInNodeWithZeroWeightedRouting(next.getNodeId())) {
+
+            if (canFailOpen(next.getShardId())) {
+                logFailOpen(next.getShardId());
+                break;
+            }
+            next = shardIt.nextOrNull();
+        }
+        return next;
+    }
+
+    /**
+     * This function returns next shard copy to retry search request in case of failure from previous copy returned
+     * by the iterator. It has the logic to fail open ie request shard copies present in nodes with weighted shard
+     * routing weight set to zero
+     *
+     * @param shardsIt Shard Iterator containing order in which shard copies for a shard need to be requested
+     * @return the next shard copy
+     */
+    public ShardRouting findNext(final ShardsIterator shardsIt) {
+        ShardRouting next = shardsIt.nextOrNull();
+
+        while (next != null && shardInNodeWithZeroWeightedRouting(next.currentNodeId())) {
+            if (canFailOpen(next.shardId())) {
+                logFailOpen(next.shardId());
+                break;
+            }
+            next = shardsIt.nextOrNull();
+        }
+        return next;
+    }
+
+    private void logFailOpen(ShardId shardID) {
+        logger.info(() -> new ParameterizedMessage("{}: Fail open executed", shardID));
+    }
```

Comment:
nit : do we need a function here or can inline it directly in `findNext `

Reply:
since this is used at multiple place, I think its better to make a function